python/shark_turbine/dynamo/tensor.py

```diff
@@ -372,6 +466,132 @@ def _rand(*args, dtype=None):
+        if isinstance(py_arg, (int, float)):
+            ret_val = DeviceTensor.from_torch(torch.tensor(py_arg))
+        devices_set.add(ret_val.device)
+        arg_shape_dtype_encode.append(str(ret_val.shape) + str(ret_val.dtype))
+        py_args.append(ret_val)
+
+    # Check if turbine device exist. If doesn't run regular fn.
+    if not any_turbine_tensor:
+        super_fn(*py_args, **kwargs)
+
+    # Do not support interop between Turbine and other devices.
+    if len(devices_set) > 1:
+        raise ValueError("Turbine do not support mixed device!")
+    cur_device = py_args[0].device
+    # Get a unique encoding to identify computation/dispatch using opCode, input shapes, and dtypes.
+    if isinstance(src_op, torch._ops.OpOverload):
+        src_op_name = src_op.name()
+    elif isinstance(src_op, BuiltinFunctionType):
+        src_op_name = src_op.__name__
+    else:
+        raise ValueError("Expected srcOp to be torchOp or builtinFn.")
+    compute_id_encode = src_op_name + "".join(arg_shape_dtype_encode)
+    compute_hash = hash(compute_id_encode)
+    if compute_hash in TurbineMode.CACHED_IMPLEMENTATIONS:
+        # TODO: Handle multiple output.
+        exec_res = TurbineMode.CACHED_IMPLEMENTATIONS[compute_hash](*py_args, **kwargs)[
+            0
+        ]
+        res_buf = DeviceTensor._from_buffer(
+            exec_res.buffer, exec_res.size, exec_res.dtype, cur_device, exec_res.signal
+        )
+        return res_buf
+
+    # Preprocess func and generate into FX.
+    flat_pytorch_args = [py_arg._to_meta_tensor() for py_arg in py_args]
+
+    # TODO: Replace all the below with torch.compile, although currently seems like
+    #       the problem lies in it will try to generate DeviceTensor, but it would be missing
+    #       _storage and causes error.
+    def func_src_op(*args, **kwargs):
+        return src_op(*args, **kwargs)
+
+    exported_f = dynamo.export(
+        func_src_op,
+        aten_graph=True,
+        decomposition_table={},
+        constraints={},
+        assume_static_by_default=True,
+    )
+    gm, guards = exported_f(*flat_pytorch_args)
+
+    # Setup mlir compilation pipeline.
+    session = Session()
+    session.set_flags(*DEFAULT_COMPILER_FLAGS)
+    session.set_flags("--iree-hal-target-backends=llvm-cpu")
+    context = session.context
+
+    # Generate MLIR from FX.
+    importer = FxImporter(context=context)
+    module = importer.module
+    inv = session.invocation()
+    # TODO: Should capture diagnostics.
+    inv.enable_console_diagnostics()
+    inv.import_module(module.operation)
+    importer.import_graph_module(gm)
+    with context:
+        pm = PassManager.parse("builtin.module(torch-to-iree)")
+        pm.run(module.operation)
+
+    # Compile MLIR to vmfb.
+    inv.execute()
+    output = Output.open_membuffer()
+    inv.output_vm_bytecode(output)
+
+    # Map VMFB to buffer.
+    device_state = _get_device_state()
+    vmfb_module = VmModule.copy_buffer(
```

Comment:
Use `wrap_buffer`. The `Output` generates a sufficiently aligned buffer to be used directly. On Linux, it is better than that and doesn't ever even touch the heap (just consumes some space in the swapfile).

Reply:
done